### PR TITLE
docs: Link memoization and work-avoidance

### LINKED ISSUE #10769: Memoization doesn't work in templates that are steps
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

`memoize` configuration added to a `steps` template doesn't have any effect. No ConfigMap is created at all.

Memoization status is always like this after multiple submits:

```
      memoizationStatus:
        cacheName: cache-this-cm-doesnt-exist
        hit: false
        key: hello-10
```

It prevents us from caching larger reusable chunks of work exposed via Workflow Templates. For example, building images is a multiple step effort – we have to build an app, login to container registry, build Dockerfile and push image. Entire thing could be cached as a whole, but it doesn't work.

There are few workarounds we considered. We're not happy with neither of them. 

1. Add `memoize` to all of the steps. It's hard to maintain and it's error-prone and sometimes incorrect. For example we we shouldn't cache "container registry login" step, because token it generates has short time to live, but also we don't need it at all if the image is already built (and it's waste of time and resources to login just for nothing).
2. Merge all steps into one, which is bad for code reusability and it's harder to maintain special images with all the tools installed (they're also bigger, so slower).
3. Add memoize to the slowest templates only and run these faster ones unnecessarily.

Image building is just an example. We have more use cases for this.

Is there a way to fix or work around this? Thanks!

Similar issue was reported for dag: https://github.com/argoproj/argo-workflows/issues/10426

### Version

v3.4.5

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: testing-memoization-mini-
  annotations:
    workflows.argoproj.io/description: Test memoization
spec:
  entrypoint: main
  templates:
    - name: main
      steps:
        - - name: echo-and-wait-step
            template: echo-and-wait
            arguments:
              parameters:
                - name: time
                  value: 10
                - name: message
                  value: "hello"
      memoize:
        key: "hello-10"
        cache:
          configMap:
            name: cache-this-cm-doesnt-exist

    - name: echo-and-wait
      inputs:
        parameters:
          - name: message
          - name: time
      script:
        image: alpine:3.17.2
        command: [sh]
        source: |
          set -e
          date
          echo {{inputs.parameters.message}}
          sleep {{inputs.parameters.time}}
```

### Logs from the workflow controller

```text
time="2023-03-28T18:19:27.757Z" level=info msg="Processing workflow" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:27.762Z" level=info msg="Updated phase  -> Running" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:27.767Z" level=info msg="Steps node testing-memoization-mini-d2k2z initialized Pending" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:27.767Z" level=info msg="StepGroup node testing-memoization-mini-d2k2z-1462015324 initialized Running" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:27.767Z" level=info msg="Pod node testing-memoization-mini-d2k2z-2612586120 initialized Pending" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:27.792Z" level=info msg="Created pod: testing-memoization-mini-d2k2z[0].echo-and-wait-step (testing-memoization-mini-d2k2z-echo-and-wait-2612586120)" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:27.792Z" level=info msg="Workflow step group node testing-memoization-mini-d2k2z-1462015324 not yet completed" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:27.792Z" level=info msg="TaskSet Reconciliation" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:27.792Z" level=info msg=reconcileAgentPod namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:27.801Z" level=info msg="Workflow update successful" namespace=ci-workflows phase=Running resourceVersion=273753196 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:37.799Z" level=info msg="Processing workflow" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:37.799Z" level=info msg="Task-result reconciliation" namespace=ci-workflows numObjs=0 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:37.799Z" level=info msg="node changed" namespace=ci-workflows new.message=PodInitializing new.phase=Pending new.progress=0/1 nodeID=testing-memoization-mini-d2k2z-2612586120 old.message= old.phase=Pending old.progress=0/1 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:37.800Z" level=info msg="Workflow step group node testing-memoization-mini-d2k2z-1462015324 not yet completed" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:37.800Z" level=info msg="TaskSet Reconciliation" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:37.800Z" level=info msg=reconcileAgentPod namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:37.811Z" level=info msg="Workflow update successful" namespace=ci-workflows phase=Running resourceVersion=273753277 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:47.812Z" level=info msg="Processing workflow" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:47.813Z" level=info msg="Task-result reconciliation" namespace=ci-workflows numObjs=0 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:47.813Z" level=info msg="node unchanged" namespace=ci-workflows nodeID=testing-memoization-mini-d2k2z-2612586120 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:47.813Z" level=info msg="Workflow step group node testing-memoization-mini-d2k2z-1462015324 not yet completed" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:47.813Z" level=info msg="TaskSet Reconciliation" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:47.813Z" level=info msg=reconcileAgentPod namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:57.906Z" level=info msg="Processing workflow" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:57.906Z" level=info msg="Task-result reconciliation" namespace=ci-workflows numObjs=0 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:57.907Z" level=info msg="node changed" namespace=ci-workflows new.message= new.phase=Running new.progress=0/1 nodeID=testing-memoization-mini-d2k2z-2612586120 old.message=PodInitializing old.phase=Pending old.progress=0/1 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:57.907Z" level=info msg="Workflow step group node testing-memoization-mini-d2k2z-1462015324 not yet completed" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:57.907Z" level=info msg="TaskSet Reconciliation" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:57.907Z" level=info msg=reconcileAgentPod namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:19:57.917Z" level=info msg="Workflow update successful" namespace=ci-workflows phase=Running resourceVersion=273753440 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.944Z" level=info msg="Processing workflow" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.945Z" level=info msg="Task-result reconciliation" namespace=ci-workflows numObjs=1 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.945Z" level=info msg="task-result changed" namespace=ci-workflows nodeID=testing-memoization-mini-d2k2z-2612586120 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.945Z" level=info msg="node changed" namespace=ci-workflows new.message= new.phase=Succeeded new.progress=0/1 nodeID=testing-memoization-mini-d2k2z-2612586120 old.message= old.phase=Running old.progress=0/1 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.945Z" level=info msg="Step group node testing-memoization-mini-d2k2z-1462015324 successful" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.945Z" level=info msg="node testing-memoization-mini-d2k2z-1462015324 phase Running -> Succeeded" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.945Z" level=info msg="node testing-memoization-mini-d2k2z-1462015324 finished: 2023-03-28 18:20:08.945950054 +0000 UTC" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.945Z" level=info msg="Outbound nodes of testing-memoization-mini-d2k2z-2612586120 is [testing-memoization-mini-d2k2z-2612586120]" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.945Z" level=info msg="Outbound nodes of testing-memoization-mini-d2k2z is [testing-memoization-mini-d2k2z-2612586120]" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.945Z" level=info msg="node testing-memoization-mini-d2k2z phase Pending -> Succeeded" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.945Z" level=info msg="node testing-memoization-mini-d2k2z finished: 2023-03-28 18:20:08.945993045 +0000 UTC" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.946Z" level=info msg="Checking daemoned children of testing-memoization-mini-d2k2z" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.946Z" level=info msg="TaskSet Reconciliation" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.946Z" level=info msg=reconcileAgentPod namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.946Z" level=info msg="Updated phase Running -> Succeeded" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.946Z" level=info msg="Marking workflow completed" namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.946Z" level=info msg="Checking daemoned children of " namespace=ci-workflows workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.951Z" level=info msg="cleaning up pod" action=deletePod key=ci-workflows/testing-memoization-mini-d2k2z-1340600742-agent/deletePod
time="2023-03-28T18:20:08.957Z" level=info msg="Workflow update successful" namespace=ci-workflows phase=Succeeded resourceVersion=273753515 workflow=testing-memoization-mini-d2k2z
time="2023-03-28T18:20:08.958Z" level=info msg="Queueing Succeeded workflow ci-workflows/testing-memoization-mini-d2k2z for delete in 4h0m0s due to TTL"
time="2023-03-28T18:20:08.978Z" level=info msg="cleaning up pod" action=labelPodCompleted key=ci-workflows/testing-memoization-mini-d2k2z-echo-and-wait-2612586120/labelPodCompleted
```

### Logs from in your workflow's wait container

```text
time="2023-03-28T18:19:48.072Z" level=info msg="capturing logs" argo=true
Tue Mar 28 18:19:48 UTC 2023
hello
time="2023-03-28T18:19:58.078Z" level=info msg="sub-process exited" argo=true error="<nil>"
```

### PR DESCRIPTION
**Pull request by Joibel**
Despite #10769 and #10426 both having examples of memoization not working with the examples having no output, no-one has picked up on this.

To address this improve the documentation for memoization and work-avoidance, linking the two ideas and pointing people who want to skip steps towards work-avoidance unless they are really doing what memoization was designed to do.

Issue #10426 is problematic in that some steps get memoized when perhaps they should't, so this commit shouldn't close it.

Fixes #10769

### CODE REVIEW

File: docs/memoization.md [hunk 1]
```diff
@@ -5,12 +5,16 @@
 ## Introduction
 
 Workflows often have outputs that are expensive to compute.
-This feature reduces cost and workflow execution time by memoizing previously run steps:
+Memoization reduces cost and workflow execution time by recording the result of previously run steps:
 it stores the outputs of a template into a specified cache with a variable key.
 
+Memoization only works for steps which have outputs, if you attempt to use it on steps which do not it should not work (there are some cases where it does, but they shouldn't). It is designed for 'pure' steps, where the purpose of running the step is to calculate some outputs based upon the steps inputs, and only the inputs. Pure steps should not interact with the outside world, but workflows won't enforce this on you.
```
Comment (tico24) [inline docs/memoization.md:11]: "steps inputs" needs an appropriate apostrophe.

File: docs/work-avoidance.md [hunk 2]
```diff
@@ -2,7 +2,11 @@
 
 > v2.9 and after
```
Comment (tico24) [inline docs/work-avoidance.md:3]: did we ever work out why this is a v2.9 feature?

**Review by tico24 — APPROVED** [review]
nits

File: docs/work-avoidance.md (hunk 2 above)
Comment (Joibel) [reply docs/work-avoidance.md:3]: No idea, unless PVCs were introduced then.

**Comment by moleskin-smile** [comment]
Thanks for clarifying on intended usage of Argo Workflows features.

Since an issue I've submitted triggered this change, let me describe our use case.

We're using Argo Workflows for CI and we're using memoization for work avoidance and I think it's valid usage. We don't want to run tests again for the same sets of inputs. So our expensive to compute output is... the information that tests pass for the given set of inputs. And we're happy that hooks run for memoized steps, because what we really need is green GitHub status check.

"Work avoidance" is not really Argo Workflows feature. It's just a technique that could be used to implement caching anywhere (which we also use with Argo in some places). But in order to not "abuse" memoization (replace memoization with "work avoidance"), we would have to repeat the same (part of) shell script in every template we want to cache, add input and output artifacts, handle artifact lifecycle etc. That's a lot of repeated YAML and shell code. And also pods would still be scheduled, run, they would fetch all input artifacts (cloning repositories for example) and would waste requested resources, increasing costs by a lot (especially for templates requesting a lot of cpu and memory). Why do that if we could just say "don't run this again for this set of inputs" which is exactly what memoization does?

Why not treat the success of template run part of its outputs and allow to memoize templates without outputs?

Memoization works really good when it works (when we don't run into 1MB limit of configmaps or [this](https://github.com/argoproj/argo-workflows/issues/11219)). It makes cached steps to be instant and makes Argo Workflows possible with monorepos.

(Let me know if this is not the right place to raise this concern).

**Comment by mnarozny** [comment]
I would also argue that this default approach is a bit odd to me. The memoization feature seems extremely powerful and restricting it to only steps with output seems counterintuitive when in a lot of cases it is the outcome with a set of parameters that matters, not necessarily the pure output. So if I memoized on my inputs and knew that the steps will produce the same output every time, I was expecting it not to have to run but give me a stamp of approval on the top level. 

I mean I could put all the steps inside into a single template and memoize it but it seems counter intuitive when I can divide it into smaller reusable templates and then mix and match them and memoize the whole thing because the outcome of that whole template will never change for the same set of inputs.

**Comment by Joibel** [comment]
I understand that people are unhappy that memoization doesn't work without outputs, and would like it to work.

This PR doesn't change any behaviour, it just tries to document the current behaviour, which even in the event of a PR to change the behaviour of memoziation needs documenting for users prior to that change.

@terrytangyuan, I'd like you to consider merging this as is unless you feel it's factually incorrect.

I have raised #11280 as a proposal to implement memoization as @moleskin-smile and @mnarozny would like it to work to see what the feelings of the community are.

**Review by terrytangyuan — APPROVED** [review]
Thanks!

File: docs/work-avoidance.md [hunk 3]
```diff
@@ -2,7 +2,11 @@
...
 
-You can make workflows faster and more robust by employing **work avoidance**. A workflow that utilizes this is simply a workflow containing steps that do not run if the work has already been done. This simplest way to do this is to use **marker files**.
+You can make workflows faster and more robust by employing **work avoidance**. A workflow that utilizes this is simply a workflow containing steps that do not run if the work has already been done.
+
+This technique is similar to [memoization](memoization.md) but they have distinct use cases. Work avoidance is totally in your control and you make the decisions as to have to skip the work. [Memoization](memoization.md) is a feature of Argo Workflows to automatically skip steps which generate outputs - it is designed
```
Comment (agilgur5) [inline docs/work-avoidance.md:7]: The last sentence here was incomplete; it ends with "it is designed". Not sure how it was intended to be written

Comment (Joibel) [reply docs/work-avoidance.md:7]: Oh, erm... yeah, I'll finish that off. Thanks @agilgur5

Comment (agilgur5) [reply docs/work-avoidance.md:7]: Fixed in #11357 . Thanks!

**Pull request merged**